### PR TITLE
Ocd 2275 - Migrate Summary Statistics

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: chpl/chpl-resources/src/main/resources-staging/jobs.xml [hunk 1]
```diff
@@ -105,5 +140,17 @@
                 <cron-expression>0 1 1 ? JAN,APR,JUL,OCT 7#1</cron-expression> <!-- At 0101 UTC on the first Saturday of the quarter -->
             </cron>
         </trigger>
+        <trigger>
+            <cron>
+                <name>generateSummaryStatistics</name>
+                <group>summaryStatisticsTrigger</group>
+                <job-name>summaryStatisticsCreator</job-name>
+                <job-group>systemJobs</job-group>
+                <misfire-instruction>MISFIRE_INSTRUCTION_FIRE_ONCE_NOW</misfire-instruction>
```
Comment (andlar) [inline chpl/chpl-resources/src/main/resources-staging/jobs.xml]: Why this misfire instruction?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java [hunk 2]
```diff
@@ -97,6 +91,9 @@ public static void main(final String[] args) throws Exception {
             while (endDate.compareTo(endDateCal.getTime()) >= 0) {
                 LOGGER.info("Getting csvRecord for start date " + startDateCal.getTime().toString() + " end date "
                         + endDateCal.getTime().toString());
+                System.out.println("Getting csvRecord for start date "
+                        + startDateCal.getTime().toString() + " end date "
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java]: Why is this here? Why not use the LOGGER?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java [hunk 3]
```diff
@@ -121,8 +121,10 @@ public static void main(final String[] args) throws Exception {
             files.add(csvFile);
         }
         String htmlMessage = summaryStats.createHtmlMessage(emailBodyStats);
+        LOGGER.info(htmlMessage);
 
         // send the email
+        /*
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java]: Was this moved to the quartz job? If so, it shouldn't be here

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/schedule/ChplTrigger.java [hunk 4]
```diff
@@ -44,8 +39,10 @@
     public ChplTrigger(final CronTrigger quartzTrigger) {
         this.name = quartzTrigger.getKey().getName();
         this.group = quartzTrigger.getKey().getGroup();
-        this.jobName = quartzTrigger.getJobKey().getName();
-        this.jobGroup = quartzTrigger.getJobKey().getName();
+        this.job = new ChplJob();
+        this.job.setDescription("");
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/schedule/ChplTrigger.java]: Shouldn't the description be set to something?

File: chpl/chpl-service/src/main/resources/log4j2-app.xml [hunk 5]
```diff
@@ -17,6 +17,9 @@
                     interval="1" modulate="true" />
             </Policies>
         </RollingFile>
+        <Console name="Console" target="SYSTEM_OUT">
+            <PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n"/>
+        </Console>
```
Comment (andlar) [inline chpl/chpl-service/src/main/resources/log4j2-app.xml]: Are you changing the resources-dev, -staging, and -production log4j files also?

File: chpl/chpl-resources/src/main/resources-dev/jobs.xml [hunk 6]
```diff
@@ -4,40 +4,59 @@
...
+                <entry>
+                    <key>frequency</key>
+                    <value>DAILY</value>
+                </entry>
+                
+            </job-data-map>
         </job>
-        <trigger>
-            <cron>
-                <name>alarned@ainq_com</name>
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources-dev/jobs.xml]: Because it is no longer valid, due to the name change with the associated job.

File: chpl/chpl-resources/src/main/resources-dev/jobs.xml [hunk 7]
```diff
@@ -75,13 +94,29 @@
...
             </job-data-map>
         </job>
+        
+        <job>
+            <name>summaryStatisticsCreator</name>
+            <group>systemJobs</group>
+            <description>Generates the summary statistics report</description>
+            <job-class>gov.healthit.chpl.scheduler.job.summarystatistics.SummaryStatisticsCreatorJob</job-class>
+            <durability>true</durability>
+            <recover>true</recover>
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources-dev/jobs.xml]: Doesn't matter to me - I figured if a job that collected data was not executed due to the a system fault, we should re-try the job.  Otherwise everyone who has a trigger to the associated email job will receive the same data they got previously.  I'll make the change.

File: chpl/chpl-resources/src/main/resources-dev/jobs.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources-dev/jobs.xml]: I don't think it is valid now, since the associated job's name has changed.  Shouldn't it be triggered with a trigger created through the app?

File: PENDING_PENDING_RELEASE_NOTES.md [hunk 8]
```diff
@@ -5,5 +5,8 @@ _Date TBD_
 ### New Features
 * Add nonconformity charts statistics endpoint 
 * Identify questionable activity based on the confirmed date of a listing rather than the certification date.
+* Added new Quartz job to gather and store Suammry Statistics
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply PENDING_PENDING_RELEASE_NOTES.md]: Addressed

File: chpl/chpl-resources/src/main/resources-dev/jobs.xml [hunk 9]
```diff
@@ -105,5 +140,17 @@
...
             </cron>
         </trigger>
+        <trigger>
+            <cron>
+                <name>generateSummaryStatistics</name>
+                <group>summaryStatisticsTrigger</group>
+                <job-name>summaryStatisticsCreator</job-name>
+                <job-group>systemJobs</job-group>
+                <misfire-instruction>MISFIRE_INSTRUCTION_FIRE_ONCE_NOW</misfire-instruction>
+                <cron-expression>0 0 0,3 ? * * *</cron-expression>
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources-dev/jobs.xml]: Addressed

File: chpl/chpl-resources/src/main/resources-staging/jobs.xml [hunk 10]
```diff
@@ -4,40 +4,59 @@
...
+                <entry>
+                    <key>frequency</key>
+                    <value>DAILY</value>
+                </entry>
+                
+            </job-data-map>
         </job>
-        <trigger>
-            <cron>
-                <name>amore@ainq_com</name>
```
[the remark this answers is not in the document]
Comment (tmy1313) [reply chpl/chpl-resources/src/main/resources-staging/jobs.xml]: See above.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java (hunk 2 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java]: File should never have been changed.  It has been reverted.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java (hunk 3 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/app/statistics/SummaryStatistics.java]: File should never have been changed. It has been reverted.

File: chpl/chpl-service/src/main/resources/log4j2-app.xml (hunk 5 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/resources/log4j2-app.xml]: No - these changes were just made to make my local debugging easier.  There was no intention to change how logging works in any other environment.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/schedule/ChplTrigger.java (hunk 4 above)
Comment (tmy1313) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/schedule/ChplTrigger.java]: In this situation, we don't have access to the JobDetails object.  A CronTrigger object only has a reference to the JobKey.  We could use the ChplSchedulerReference singleton, to get the JobDetail based on the JobKEy.  Do you think this class should a reference to ChplSchedulerReference?

File: chpl/chpl-resources/src/main/resources-dev/jobs.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (andlar) [reply chpl/chpl-resources/src/main/resources-dev/jobs.xml]: The old triggers will be deleted during the ingest of the new jobs.xml file, but there should be triggers associated with the new job in the -dev and -staging system, as a politeness thing.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/schedule/ChplTrigger.java (hunk 4 above)
Comment (andlar) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/domain/schedule/ChplTrigger.java]: I guess it doesn't really matter in any case as we don't use the Description of a trigger anywhere yet

**Comment by andlar** [comment]
One of the things we should do when migrating from an app to Quartz is remove the old app, including whatever shell script was needed to start it. The end goal is to get rid of all of the /app/job files. I don't see that removal in this PR yet

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/dao/statistics/SummaryStatisticsDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+import gov.healthit.chpl.auth.Util;
+import gov.healthit.chpl.dao.impl.BaseDAOImpl;
+import gov.healthit.chpl.entity.SummaryStatisticsEntity;
+import gov.healthit.chpl.exception.EntityCreationException;
+import gov.healthit.chpl.exception.EntityRetrievalException;
+
+@Repository("summaryStatisticsDAO")
+public class SummaryStatisticsDAOImpl extends BaseDAOImpl implements SummaryStatisticsDAO {
+    private static final long MODIFIED_USER_ID = -3L;
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/dao/statistics/SummaryStatisticsDAOImpl.java]: There are a number of DAO classes that have the same MODIFIED_USER_ID defined. Could you move that into some common util-type place as a public static variable?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/SummaryStatisticsCreatorJob.java [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+                    + endDateCal.getTime().toString());
+            System.out.println("Finished getting csvRecord for start date "
+                    + startDateCal.getTime().toString() + " end date "
+                    + endDateCal.getTime().toString());
+            startDateCal.add(Calendar.DATE, numDaysInPeriod);
+            endDateCal.setTime(startDateCal.getTime());
+            endDateCal.add(Calendar.DATE, numDaysInPeriod);
+        }
+        LOGGER.info("Finished getting statistics");
+        System.out.println("Finished getting statistics");
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/summarystatistics/SummaryStatisticsCreatorJob.java]: Could you make sure the System.out.printlns are removed? I think the LOGGER line should be sufficient but if you weren't able to see the logs and needed system outs for some reason then we should revisit the logging settings.
